Libraries/Opc.Ua.Server/Subscription/MonitoredItem/Queue/EventMonitoredItemQueue.cs

```diff
@@ -0,0 +1,185 @@
+        /// </summary>
+        /// <param name="queueSize">The new queue size.</param>
+        /// <param name="discardOldest">if true remove oldest entries from the queue when the queue size decreases, else remove newest</param>
+        void SetQueueSize(uint queueSize, bool discardOldest);
+
+        /// <summary>
+        /// Checks the last 1k queue entries if the event is already in there
+        /// used to detect duplicate instances of the same event being reported via multiple paths.
+        /// </summary>
+        /// <param name="instance">the event to chack for duplicates</param>
+        /// <returns>true if event already in queue</returns>
+        bool IsEventContainedInQueue(IFilterTarget instance);
+
+
+        /// <summary>
+        /// Adds the value to the queue.
+        /// </summary>
+        /// <param name="value">The value to queue.</param>
+        void Enqueue(EventFieldList value);
+
+        /// <summary>
+        /// Dequeue the oldest value in the queue.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <returns>True if a value was found. False if the queue is empty.</returns>
+        bool Dequeue(out EventFieldList value);
+    }
+    /// <summary>
+    /// Provides a queue for events.
+    /// </summary>
+    public class EventMonitoredItemQueue : IEventMonitoredItemQueue
+    {
+        /// <summary>
+        /// Creates an empty queue.
+        /// </summary>
+        public EventMonitoredItemQueue(bool createDurable)
+        {
+            if (createDurable)
+            {
+                Utils.LogError("EventMonitoredItemQueue does not support durable queues, please provide full implementation of IDurableMonitoredItemQueue using Server.CreateDurableMonitoredItemQueueFactory to supply own factory");
+                throw new ArgumentException("DataChangeMonitoredItemQueue does not support durable Queues", nameof(createDurable));
+            }
+            m_events = new List<EventFieldList>();
+            QueueSize = 0;
+        }
+
+        #region Public Methods
+        /// <inheritdoc/>
+        public virtual bool IsDurable => false;
+
+        /// <inheritdoc/>
+        public uint QueueSize { get; private set; }
+
+        /// <inheritdoc/>
+        public int ItemsInQueue => m_events.Count;
+
+        /// <inheritdoc/>
+        public bool Dequeue(out EventFieldList value)
+        {
+            value = null;
+            if (m_events.Any())
+            {
+                value = m_events.First();
+                m_events.RemoveAt(0);
+                return true;
+            }
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            //Only needed for unmanaged resources
+        }
+
+        /// <inheritdoc/>
+        public void Enqueue(EventFieldList value)
+        {
+            if (m_events.Count == QueueSize && !Dequeue(out var _))
```

Comment:
If first condition checks if the queue is full then why !Deque(out var _) ? Doesn't second condition return true only if queue is empty ?